File: alligator/src/main/java/me/aartikov/alligator/commands/BackCommand.java
package me.aartikov.alligator.commands;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import me.aartikov.alligator.NavigationContext;
import me.aartikov.alligator.ScreenResult;
import me.aartikov.alligator.animations.AnimationData;
import me.aartikov.alligator.exceptions.NavigationException;


/**
 * Command implementation for {@code goBack} method of {@link me.aartikov.alligator.AndroidNavigator}.
 */
public class BackCommand implements Command {
	@Nullable
	private ScreenResult mScreenResult;
	@Nullable
	private AnimationData mAnimationData;

	public BackCommand(@Nullable ScreenResult screenResult, @Nullable AnimationData animationData) {
		mScreenResult = screenResult;
		mAnimationData = animationData;
	}

	@Override
	public boolean execute(@NonNull NavigationContext navigationContext) throws NavigationException {
		if (navigationContext.getDialogFragmentNavigator().canGoBack()) {
			navigationContext.getDialogFragmentNavigator().goBack(mScreenResult);
			return true;
		} else if (navigationContext.getFragmentNavigator() != null && navigationContext.getFragmentNavigator().canGoBack()) {
			navigationContext.getFragmentNavigator().goBack(mScreenResult, mAnimationData);
			return true;
		} else {
			navigationContext.getActivityNavigator().goBack(mScreenResult, mAnimationData);
			return false;
		}
	}
}

File: alligator/src/main/java/me/aartikov/alligator/commands/BackCommand.kt
package me.aartikov.alligator.commands

import me.aartikov.alligator.NavigationContext
import me.aartikov.alligator.ScreenResult
import me.aartikov.alligator.animations.AnimationData
import me.aartikov.alligator.exceptions.NavigationException

/**
 * Command implementation for `goBack` method of [me.aartikov.alligator.AndroidNavigator].
 */
class BackCommand(
    private val mScreenResult: ScreenResult?,
    private val mAnimationData: AnimationData?
) : Command {
    @Throws(NavigationException::class)
    override fun execute(navigationContext: NavigationContext): Boolean {
        return if (navigationContext.dialogFragmentNavigator.canGoBack()) {
            navigationContext.dialogFragmentNavigator.goBack(mScreenResult)
            true
        } else if (navigationContext.fragmentNavigator != null && navigationContext.fragmentNavigator!!
                .canGoBack()
        ) {
            navigationContext.fragmentNavigator!!.goBack(mScreenResult, mAnimationData)
            true
        } else if (navigationContext.flowFragmentNavigator != null && navigationContext.flowFragmentNavigator!!
                .canGoBack()
        ) {
            navigationContext.flowFragmentNavigator!!.goBack(mScreenResult, mAnimationData)
            true
        } else {
            navigationContext.activityNavigator.goBack(mScreenResult, mAnimationData)
            false
        }
    }
}